gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java

```diff
@@ -122,82 +122,22 @@
-
-      DatasetsFinder<CopyableDatasetBase> datasetFinder = DatasetUtils
-          .instantiateDatasetFinder(state.getProperties(), sourceFs, DEFAULT_DATASET_PROFILE_CLASS_KEY,
-              new EventSubmitter.Builder(this.metricContext, CopyConfiguration.COPY_PREFIX).build(), state);
-
-      IterableDatasetFinder<CopyableDatasetBase> iterableDatasetFinder =
-          datasetFinder instanceof IterableDatasetFinder ? (IterableDatasetFinder<CopyableDatasetBase>) datasetFinder
-              : new IterableDatasetFinderImpl<>(datasetFinder);
-
-      Iterator<CopyableDatasetBase> copyableDatasets =
-          new InterruptibleIterator<>(iterableDatasetFinder.getDatasetsIterator(), new Callable<Boolean>() {
-            @Override
-            public Boolean call()
-                throws Exception {
-              return shouldStopGeneratingWorkUnits(workUnitList);
-            }
-          });
-
-      Iterator<Callable<Void>> callableIterator =
-          Iterators.transform(copyableDatasets, new Function<CopyableDatasetBase, Callable<Void>>() {
-            @Nullable
-            @Override
-            public Callable<Void> apply(@Nullable CopyableDatasetBase copyableDataset) {
-
-              IterableCopyableDataset iterableCopyableDataset;
-              if (copyableDataset instanceof IterableCopyableDataset) {
-                iterableCopyableDataset = (IterableCopyableDataset) copyableDataset;
-              } else if (copyableDataset instanceof CopyableDataset) {
-                iterableCopyableDataset = new IterableCopyableDatasetImpl((CopyableDataset) copyableDataset);
-              } else {
-                throw new RuntimeException(String.format("Cannot process %s, can only copy %s or %s.",
-                    copyableDataset == null ? null : copyableDataset.getClass().getName(),
-                    CopyableDataset.class.getName(), IterableCopyableDataset.class.getName()));
-              }
-
-              return new DatasetWorkUnitGenerator(iterableCopyableDataset, sourceFs, targetFs, state, workUnitList,
-                  copyConfiguration, minWorkUnitWeight, watermarkGenerator);
-            }
-          });
-
-      try {
-        List<Future<Void>> futures = new IteratorExecutor<>(callableIterator,
-            state.getPropAsInt(MAX_CONCURRENT_LISTING_SERVICES, DEFAULT_MAX_CONCURRENT_LISTING_SERVICES),
-            ExecutorsUtils.newDaemonThreadFactory(Optional.of(log), Optional.of("Copy-file-listing-pool-%d")))
-            .execute();
-
-        for (Future<Void> future : futures) {
-          try {
-            future.get();
-          } catch (ExecutionException exc) {
-            log.error("Failed to get work units for dataset.", exc.getCause());
-          }
-        }
-      } catch (InterruptedException ie) {
-        log.error("Retrieval of work units was interrupted. Aborting.");
+      if(!generateWorkunitsBasedOnConfig(state, workUnitList, minWorkUnitWeight)){
         return Lists.newArrayList();
```

Comment:
nit: Collections.emptyList() is slight more efficient since it returns a immutable singleton instance.